src/Kestrel.Core/Internal/HttpsConnectionAdapter.cs

```diff
@@ -36,15 +38,34 @@ public HttpsConnectionAdapter(HttpsConnectionAdapterOptions options, ILoggerFact
                 throw new ArgumentNullException(nameof(options));
             }
 
-            if (options.ServerCertificate == null)
+            // capture the certificate now so it can't be switched after validation
+            _serverCertificate = options.ServerCertificate;
+            var selector = options.ServerCertificateSelector;
+            if (_serverCertificate == null && selector == null)
             {
-                throw new ArgumentException(CoreStrings.ServiceCertificateRequired, nameof(options));
+                throw new ArgumentException(CoreStrings.ServerCertificateRequired, nameof(options));
             }
 
-            // capture the certificate now so it can be switched after validation
-            _serverCertificate = options.ServerCertificate;
-
-            EnsureCertificateIsAllowedForServerAuth(_serverCertificate);
+            // If a selector is provided then ignore the cert, it may be a default cert.
+            if (selector != null)
+            {
+                // SslStream doesn't allow both.
+                _serverCertificate = null;
+                // Adapt to the SslStream signature
+                _serverCertificateSelector = (sender, name) =>
+                {
+                    var cert = selector(name);
+                    if (cert != null)
+                    {
+                        EnsureCertificateIsAllowedForServerAuth(cert);
```

Comment:
@blowdart it's potentially quite expensive to validate EKUs this on every new connection. Can we exclude that for the advanced SNI scenario?

Reply:
Cache the result based on cert thumbprint?

Reply:
@davidfowl what are the odds you'd allow registering and injecting a MemoryCache here? 😁

Reply:
Kestrel.Core shouldn't depend on the memory cache.